api/src/main/java/org/openmrs/api/db/hibernate/PatientSearchCriteria.java

```diff
@@ -230,135 +246,151 @@ PatientSearchMode getSearchMode(String name, String identifier, List<PatientIden
 				}
 				// if the regex is empty, default to a simple "like" search or if
 				// we're in hsql world, also only do the simple like search (because
 				// hsql doesn't know how to deal with 'regexp'
 				else if ("".equals(regex) || HibernateUtil.isHSQLDialect(sessionFactory)) {
-					conjunction.add(getCriterionForSimpleSearch(identifier, adminService));
+					predicates.add(getPredicateForSimpleSearch(cb, idsJoin, identifier, adminService));
 				}
 				// if the regex is present, search on that
 				else {
 					regex = replaceSearchString(regex, identifier);
-					conjunction.add(Restrictions.sqlRestriction("identifier regexp ?", regex, StringType.INSTANCE));
+					predicates.add(cb.isTrue(cb.function("regexp", Boolean.class, idsJoin.get("identifier"),
+						cb.literal(regex))));
 				}
 			}
 		}
-		
+
 		// do the type restriction
 		if (!CollectionUtils.isEmpty(identifierTypes)) {
-			criteria.add(Restrictions.in("ids.identifierType", identifierTypes));
+			predicates.add(idsJoin.get("identifierType").in(identifierTypes));
 		}
-		
-		return conjunction;
+
+		return cb.and(predicates.toArray(predicates.toArray(new Predicate[]{})));
 	}
-	
+
 	/**
 	 * Utility method to add identifier expression to an existing criteria
 	 *
+	 * @param cb        the CriteriaBuilder to build the criteria
+	 * @param idsJoin   the join from Patient to PatientIdentifier
 	 * @param identifier
 	 * @param identifierTypes
 	 * @param matchIdentifierExactly
 	 */
-	private Criterion prepareCriterionForIdentifier(String identifier, List<PatientIdentifierType> identifierTypes,
-	        boolean matchIdentifierExactly) {
-		return prepareCriterionForIdentifier(identifier, identifierTypes, matchIdentifierExactly, false);
+	private Predicate preparePredicateForIdentifier(CriteriaBuilder cb, Join<Patient, PatientIdentifier> idsJoin,
+	        String identifier, List<PatientIdentifierType> identifierTypes, boolean matchIdentifierExactly) {
+		return preparePredicateForIdentifier(cb, idsJoin, identifier, identifierTypes, matchIdentifierExactly, false);
 	}
-	
+
 	/**
 	 * Utility method to add prefix and suffix like expression
 	 *
+	 * @param cb        the CriteriaBuilder to build the criteria
+	 * @param idsJoin   the join from Patient to PatientIdentifier
 	 * @param identifier
 	 * @param adminService
 	 */
-	private Criterion getCriterionForSimpleSearch(String identifier, AdministrationService adminService) {
+	private Predicate getPredicateForSimpleSearch(CriteriaBuilder cb, Join<Patient, PatientIdentifier> idsJoin,
+	        String identifier, AdministrationService adminService) {
 		String prefix = adminService.getGlobalProperty(OpenmrsConstants.GLOBAL_PROPERTY_PATIENT_IDENTIFIER_PREFIX, "");
 		String suffix = adminService.getGlobalProperty(OpenmrsConstants.GLOBAL_PROPERTY_PATIENT_IDENTIFIER_SUFFIX, "");
-		return Restrictions.ilike("ids.identifier", prefix + identifier + suffix);
+		String matchPattern = prefix + identifier.toLowerCase() + suffix;
```

Comment:
Shouldn't we call `toLowerCase` for the entire `matchPattern`? That is, to ensure that `prefix` and `suffix` are also handled in a case insensitive manner?

Reply:
Yes good point, fixed this